### PR TITLE
[ADT] Add TrieRawHashMap

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Implement TrieRawHashMap can be used to store object with its associated hash. User needs to supply a strong hashing function to guarantee the uniqueness of the hash of the objects to be inserted. A hash collision is not supported and will lead to error or failed to insert.

TrieRawHashMap is thread-safe and lock-free and can be used as foundation data structure to implement a content addressible storage. TrieRawHashMap owns the data stored in it and is designed to be:
* Fast to lookup.
* Fast to "insert" if the data has already been inserted.
* Can be used without lock and doesn't require any knowledge of the participating threads or extra coordination between threads.

It is not currently designed to be used to insert unique new data with high contention, due to the limitation on the memory allocator.

### CODE REVIEW

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+  /// Linked list for ownership of tries. The pointer is owned by TrieSubtrie.
+  std::atomic<TrieSubtrie *> Next;
+
+  /// The (co-allocated) slots of the subtrie.
+  MutableArrayRef<LazyAtomicPointer<TrieNode>> Slots;
+};
+} // end namespace
+
+namespace llvm {
+template <> struct isa_impl<TrieContent, TrieNode> {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:101]: Please use `CastInfo` instead of isa_impl - the docs mention this but we're actively trying to remove isa_impl wherever we can

**Review by bzcheeseman — COMMENTED** [review]
Don't block on me, just a drive-by comment while skimming the change :) I'll take a more in-depth look when I get the chance (and plz ping if I am holding something up)

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 1 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp]: Thanks! I switch to adding `classof` in the struct so it should be handled by default CastInfo implementation and remove this code.

File: llvm/include/llvm/ADT/TrieRawHashMap.h [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+  class const_pointer : public PointerImpl<const value_type> {
+    friend class ThreadSafeTrieRawHashMap;
+
+  public:
+    const_pointer() = default;
+    const_pointer(const_pointer &&) = default;
+    const_pointer(const const_pointer &) = default;
+    const_pointer &operator=(const_pointer &&) = default;
+    const_pointer &operator=(const const_pointer &) = default;
```
Comment (dwblaikie) [inline llvm/include/llvm/ADT/TrieRawHashMap.h:283]: I think you can leave these 4 out and get them implicitly (similarly elsewhere)?

File: llvm/lib/Support/TrieHashIndexGenerator.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    } else {
+      *StartBit += *StartBit ? NumSubtrieBits : NumRootBits;
+      assert((*StartBit - NumRootBits) % NumSubtrieBits == 0);
+      Index = getIndex(Bytes, *StartBit, NumSubtrieBits);
+    }
+    return Index;
+  }
+
+  size_t hint(unsigned Index, unsigned Bit) {
+    assert(Index >= 0);
```
Comment (dwblaikie) [inline llvm/lib/Support/TrieHashIndexGenerator.h:44]: This is tautological, since `Index` is unsigned. (I'm surprised this didn't produce a warning?)

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+  //     while (!Bits.compare_exchange_weak(Old, Old | NewBit))
+  //       ;
+
+  // For debugging.
+  unsigned StartBit = 0;
+  unsigned NumBits = 0;
+  friend class llvm::ThreadSafeTrieRawHashMapBase;
+
+public:
+  /// Linked list for ownership of tries. The pointer is owned by TrieSubtrie.
```
Comment (dwblaikie) [inline llvm/lib/Support/TrieRawHashMap.cpp:96]: Maybe "This is an owning pointer"? ("The pointer is owned by TrieSubtrie" confused me a bit (it's not about what owns the pointer, but the memory it points to - and "which TrieSubtrie" - `this` or some other instance, etc))

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+  assert(NumBits < 20 && "Tries should have fewer than ~1M slots");
+  return sizeof(TrieNode *) * (1u << NumBits);
+}
+
+std::unique_ptr<TrieSubtrie> TrieSubtrie::create(size_t StartBit,
+                                                 size_t NumBits) {
+  size_t Size = sizeof(TrieSubtrie) + getTrieTailSize(StartBit, NumBits);
+  void *Memory = ::malloc(Size);
+  TrieSubtrie *S = ::new (Memory) TrieSubtrie(StartBit, NumBits);
+  return std::unique_ptr<TrieSubtrie>(S);
```
Comment (dwblaikie) [inline llvm/lib/Support/TrieRawHashMap.cpp:114]: This would produce mismatched new/delete. If the object is created with malloc+placement new, it should be destroyed with an explicit dtor call (if it's non-trivial) and a call to free - not a call to delete. I think?
(a unique_ptr with a custom deleter could be used to address this issue)

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+    TrieSubtrie *CurrentHead = nullptr;
+    // Add ownership of "S" to front of the list, so that Root -> S ->
+    // Root.Next. This works by repeatedly setting S->Next to a candidate value
+    // of Root.Next (initially nullptr), then setting Root.Next to S once the
+    // candidate matches reality.
+    while (!Root.Next.compare_exchange_weak(CurrentHead, S.get()))
+      S->Next.exchange(CurrentHead);
+
+    // Ownership transferred to subtrie.
+    return S.release();
```
Comment (dwblaikie) [inline llvm/lib/Support/TrieRawHashMap.cpp:169]: If this API is doing ownership transfer, ideally it wouldn't do so with a raw pointer - could `save` return a `std::unique_ptr` instead? (and `return S` here)

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+ThreadSafeTrieRawHashMapBase::getOrCreateImpl() {
+  if (ImplType *Impl = ImplPtr.load())
+    return *Impl;
+
+  // Create a new ImplType and store it if another thread doesn't do so first.
+  // If another thread wins this one is destroyed locally.
+  std::unique_ptr<ImplType> Impl = ImplType::create(0, NumRootBits);
+  ImplType *ExistingImpl = nullptr;
+  if (ImplPtr.compare_exchange_strong(ExistingImpl, Impl.get()))
+    return *Impl.release();
```
Comment (dwblaikie) [inline llvm/lib/Support/TrieRawHashMap.cpp:195]: That looks really problematic - releasing memory from a unique_ptr then returning a reference to it - makes the ownership very unclear, I think? (passing ownership via a reference is quite uncommon/surprising/unlikely to communicate the API needs to the caller clearly)

Ah, this is because ownership already transferred via the compare_exchange_strong? It's probably more convoluted, but maybe it'd be more clear with an explicit owning raw pointer for a short period here?
Hmm, maybe not - but maybe at least a comment on the `release` call, explaining that ownership has already been taken at this point, would be helpful?

File: llvm/unittests/ADT/TrieRawHashMapTest.cpp [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    auto &Ret = Trie.emplace(RootBits, SubtrieBits);
+    TrieRawHashMapTestHelper::setTrie(&Ret);
+    return Ret;
+  }
+
+  void destroyTrie() { Trie.reset(); }
+
+  ~SimpleTrieHashMapTest() {
+    if (Trie)
+      Trie.reset();
```
Comment (dwblaikie) [inline llvm/unittests/ADT/TrieRawHashMapTest.cpp:70]: Could call `Trie.reset()` unconditionally, I think?

File: llvm/unittests/ADT/TrieRawHashMapTest.cpp [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    TrieType::pointer Lookup = Trie.find(hash(N));
+    EXPECT_FALSE(Lookup);
+
+    Trie.insert(Lookup, TrieType::value_type(hash(N), N));
+  }
+  for (NumType N : Numbers) {
+    TrieType::pointer Lookup = Trie.find(hash(N));
+    EXPECT_TRUE(Lookup);
+    if (!Lookup)
+      continue;
```
Comment (dwblaikie) [inline llvm/unittests/ADT/TrieRawHashMapTest.cpp:214]: Might be simpler to `ASSERT_TRUE` rather than having fallbacks/etc in case this test fails?

File: llvm/unittests/ADT/TrieRawHashMapTest.cpp [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+  }
+
+  // Destroy tries. If destruction is recursive and MaxN is high enough, these
+  // will both fail.
+  destroyTrie();
+}
+
+struct NumWithDestructorT {
+  uint64_t Num;
+  ~NumWithDestructorT() {}
```
Comment (dwblaikie) [inline llvm/unittests/ADT/TrieRawHashMapTest.cpp:276]: worth doing something in the dtor & observing that it's executed?

File: llvm/lib/Support/TrieHashIndexGenerator.h (hunk 3 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieHashIndexGenerator.h:44]: Nice catch. I am surprised that it has no warning for me.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 5 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:114]: `delete` is actually overridden with `free()`. I am not proud of this but it is tricky to get this placement new but also return unique_ptr.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 6 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:169]: This is the same as next place, the ownership is transferred via a succesful compare_exchange. The owner is the `Next`, not return value. Comment explains why it calls `release`, let me clarify that.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 4 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:96]: Also this is also to avoid using recursion to destroy all tries, since the number of tries can be very very big and overflow the stack.

File: llvm/unittests/ADT/TrieRawHashMapTest.cpp (hunk 9 above)
Comment (dexonsmith) [reply llvm/unittests/ADT/TrieRawHashMapTest.cpp]: FWIW, my memory is that this code originally had `ASSERT_TRUE`, and when it failed, I got hard-to-diagnose crashes (and subsequent ADTTests didn't run). I think `EXPECT_TRUE` avoids the crashes on failure.

Comment (dexonsmith) [reply llvm/unittests/ADT/TrieRawHashMapTest.cpp]: I'm not 100% sure I'm remembering correctly; I'm pretty sure that come up somewhere in the InMemoryCAS stuff though. Probably worth trying `ASSERT_TRUE`, then temporarily forcing the condition to fail, and seeing if the test suite crashes... if not, great, leave it with ASSERT, but if it does, perhaps add a comment explaining why it's `EXPECT`.

Comment (cachemeifyoucan) [reply llvm/unittests/ADT/TrieRawHashMapTest.cpp]: EXPECT_TRUE has the benefit of checking every iterations of the loop even one of them failed so it has better coverage on failed case to provide a complete picture. On the other hand, I don't think that provide too much value when we don't expect it to fail :)

**Comment by dwblaikie** [comment]
Per https://llvm.org/docs/GitHub.html#updating-pull-requests it'd be helpful to, "When updating a pull request, you should push additional “fix up” commits to your branch instead of force pushing. This makes it easier for GitHub to track the context of previous review comments. Consider using the [built-in support for fixups](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt) in git."

(like, I can't readily tell what things changed in your most recent update - so I can see if/how different feedback was/wasn't addressed)

(I know we're all learning how to do pull requests in LLVM - I still haven't sent any of my own out... so I'm certainly no expert at making them, but I'm slowly learning what makes them easier/harder to review, at least)

File: llvm/unittests/ADT/TrieRawHashMapTest.cpp (hunk 9 above)
Comment (dexonsmith) [reply llvm/unittests/ADT/TrieRawHashMapTest.cpp]: Oh, if that's the only benefit here, I agree ASSERT is better :)

**Comment by cachemeifyoucan** [comment]
> Per https://llvm.org/docs/GitHub.html#updating-pull-requests it'd be helpful to, "When updating a pull request, you should push additional “fix up” commits to your branch instead of force pushing. This makes it easier for GitHub to track the context of previous review comments. Consider using the [built-in support for fixups](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt) in git."

Oops, I didn't realize we push out an official guideline already. Definitely will do that in the future. At the meantime, let me see if the review history can snap back to the code if I restore the branch to old stage and add fixup commit.

**Comment by cachemeifyoucan** [comment]
Ping~

File: llvm/lib/Support/TrieHashIndexGenerator.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+#ifndef LLVM_LIB_SUPPORT_TRIEHASHINDEXGENERATOR_H
+#define LLVM_LIB_SUPPORT_TRIEHASHINDEXGENERATOR_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include <optional>
+
+namespace llvm {
+
+struct IndexGenerator {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieHashIndexGenerator.h:17]: nit: Could you add comments to this struct and its functions? It also may be worth moving them out-of-line, they are maybe a little long to be fully inlined.

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+public:
+  /// Linked list for ownership of tries. The pointer is owned by TrieSubtrie.
+  std::atomic<TrieSubtrie *> Next;
+
+  /// The (co-allocated) slots of the subtrie.
+  MutableArrayRef<LazyAtomicPointer<TrieNode>> Slots;
+};
+} // end namespace
+
+static size_t getTrieTailSize(size_t StartBit, size_t NumBits) {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:105]: Can you comment this function?

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+static_assert(sizeof(TrieContent) ==
+                  ThreadSafeTrieRawHashMapBase::TrieContentBaseSize,
+              "Check header assumption!");
+
+class TrieSubtrie final : public TrieNode {
+public:
+  TrieNode *get(size_t I) const { return Slots[I].load(); }
+
+  TrieSubtrie *
+  sink(size_t I, TrieContent &Content, size_t NumSubtrieBits, size_t NewI,
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:62]: Would be great to get some comments on this, it's not super obvious from inspection what the intended behavior is :)

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+static size_t getTrieTailSize(size_t StartBit, size_t NumBits) {
+  assert(NumBits < 20 && "Tries should have fewer than ~1M slots");
+  return sizeof(TrieNode *) * (1u << NumBits);
+}
+
+std::unique_ptr<TrieSubtrie> TrieSubtrie::create(size_t StartBit,
+                                                 size_t NumBits) {
+  size_t Size = sizeof(TrieSubtrie) + getTrieTailSize(StartBit, NumBits);
+  void *Memory = ::malloc(Size);
+  TrieSubtrie *S = ::new (Memory) TrieSubtrie(StartBit, NumBits);
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:114]: Isn't operator new overloaded above? Could you use that?

I would echo David's comment below that this is a bit fussy but if this is what you have to do I believe you :)

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+
+struct ThreadSafeTrieRawHashMapBase::ImplType {
+  static std::unique_ptr<ImplType> create(size_t StartBit, size_t NumBits) {
+    size_t Size = sizeof(ImplType) + getTrieTailSize(StartBit, NumBits);
+    void *Memory = ::malloc(Size);
+    ImplType *Impl = ::new (Memory) ImplType(StartBit, NumBits);
+    return std::unique_ptr<ImplType>(Impl);
+  }
+
+  TrieSubtrie *save(std::unique_ptr<TrieSubtrie> S) {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:158]: It would be great to get some of the comments inside the function up onto the API - especially the ownership transfer bits.

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+  assert(!Hash.empty() && "Uninitialized hash");
+
+  ImplType *Impl = ImplPtr.load();
+  if (!Impl)
+    return PointerBase();
+
+  TrieSubtrie *S = &Impl->Root;
+  IndexGenerator IndexGen{NumRootBits, NumSubtrieBits, Hash};
+  size_t Index = IndexGen.next();
+  while (true) {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:216]: Is there nothing we could use as an exit condition here? What about something like `while (Index < IndexGen.max())`?

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 17]
```diff
@@ -0,0 +1,483 @@
...
+  IndexGenerator IndexGen{NumRootBits, NumSubtrieBits, Hash};
+  size_t Index;
+  if (Hint.isHint()) {
+    S = static_cast<TrieSubtrie *>(Hint.P);
+    Index = IndexGen.hint(Hint.I, Hint.B);
+  } else {
+    Index = IndexGen.next();
+  }
+
+  while (true) {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:250]: Same as above - is there any exit criteria we can put here to avoid unintentional looping?

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 18]
```diff
@@ -0,0 +1,483 @@
...
+      continue;
+    }
+
+    // Return the existing content if it's an exact match!
+    auto &ExistingContent = cast<TrieContent>(Existing);
+    if (ExistingContent.getHash() == Hash)
+      return PointerBase(ExistingContent.getValuePointer());
+
+    // Sink the existing content as long as the indexes match.
+    while (true) {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:284]: Same here - especially nested `while (true)` makes me squidgy.

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 19]
```diff
@@ -0,0 +1,483 @@
...
+
+ThreadSafeTrieRawHashMapBase::ThreadSafeTrieRawHashMapBase(
+    size_t ContentAllocSize, size_t ContentAllocAlign, size_t ContentOffset,
+    std::optional<size_t> NumRootBits, std::optional<size_t> NumSubtrieBits)
+    : ContentAllocSize(ContentAllocSize), ContentAllocAlign(ContentAllocAlign),
+      ContentOffset(ContentOffset),
+      NumRootBits(NumRootBits ? *NumRootBits : DefaultNumRootBits),
+      NumSubtrieBits(NumSubtrieBits ? *NumSubtrieBits : DefaultNumSubtrieBits),
+      ImplPtr(nullptr) {
+  assert((!NumRootBits || *NumRootBits < 20) &&
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:310]: Out of curiosity, why do we have this restriction?

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+  if (!S || !S->IsSubtrie)
+    return "";
+
+  // Find a TrieContent node which has hash stored. Depth search following the
+  // first used slot until a TrieContent node is found.
+  TrieSubtrie *Current = S;
+  TrieContent *Node = nullptr;
+  while (Current) {
+    TrieSubtrie *Next = nullptr;
+    // find first used slot in the trie.
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:416]: spelling: capital `Find`

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 21]
```diff
@@ -0,0 +1,483 @@
...
+        HashSize(HashSize), HashOffset(HashOffset) {}
+
+  static bool classof(const TrieNode *TN) { return !TN->IsSubtrie; }
+};
+
+static_assert(sizeof(TrieContent) ==
+                  ThreadSafeTrieRawHashMapBase::TrieContentBaseSize,
+              "Check header assumption!");
+
+class TrieSubtrie final : public TrieNode {
```
Comment (bzcheeseman) [inline llvm/lib/Support/TrieRawHashMap.cpp:57]: Out of curiosity, is this the only subclass of TrieNode? If so, why have the hierarchy at all? I might be missing something, but it looks like all the uses of TrieNode go through TrieSubtrie.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 14 above)
Comment (bzcheeseman) [reply llvm/lib/Support/TrieRawHashMap.cpp:114]: Actually, is https://llvm.org/doxygen/classllvm_1_1TrailingObjects.html something that might be useful here? It looks like that's what you're going for, and this might be a little cleaner?

File: llvm/lib/Support/TrieHashIndexGenerator.h (hunk 11 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieHashIndexGenerator.h:31]: Added comments for the class and APIs. Let me know if it is still hard to read for inline implementation. This is currently a private header and I can definitely move some implementation out to its own cpp file.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 14 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:115]: I guess using a TrailingObject is a cleaner solution but it does come with some downside (unless I missed how that can be used). Currently, the trailing object in `TrieSubtrie` is dynamically allocated based on parameter and I don't know how to get trailing object to work with a dynamically size object. So there are two ways to get it working:
* instead of trailing object, just put a trailing pointer there.
* for the currently implementation, the size is only affected by the configuration so technically we can template that, then we can have a static size to construct trailing object. The main problem here is that we haven't fine tuned to configuration yet and that will prevent creating a trie implementation that decides bit size of different level or you just have to create all variation of size node type ahead of time.

@dexonsmith any opinion on this?

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 16 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:227]: You will basically assertion fail if `next()` somehow get out of bound. Currently logic will return 0 which is a valid index without assertion if you feed `getIndex` with out of bound data. I guess we could make that `SIZE_MAX` if StartBit goes pass the end hash and use that as `end()`.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 19 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp]: I don't know~ @dexonsmith any insight.

I can see why it is not a reasonable choice but I don't know anything bad would happen if you force that (I don't think we will overflow, even on a 32 bit architecture). I removed it for now.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 21 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:57]: `TrieContent` is another subclass :) The node in a trie can either be a sub-trie that has children, or a trie-content that is a leaf node that store data.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 16 above)
Comment (bzcheeseman) [reply llvm/lib/Support/TrieRawHashMap.cpp]: Yeah it'd be great to not rely on assertions for correctness :)

File: llvm/lib/Support/TrieHashIndexGenerator.h (hunk 11 above)
Comment (bzcheeseman) [reply llvm/lib/Support/TrieHashIndexGenerator.h:31]: I would probably move this to a .cpp if it was me. The comments help a lot though, thanks!

File: llvm/include/llvm/ADT/TrieRawHashMap.h [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+
+  public:
+    const_pointer() = default;
+    const_pointer(const pointer &P) : const_pointer::PointerImpl(P) {}
+
+  private:
+    const_pointer(PointerBase Result) : const_pointer::PointerImpl(Result) {}
+  };
+
+  class LazyValueConstructor {
```
Comment (bzcheeseman) [inline llvm/include/llvm/ADT/TrieRawHashMap.h:272]: This could probably use a comment and/or usage example? Another "not super clear from inspection" case.

**Review by bzcheeseman — COMMENTED** [review]
Aside from the trailing object open discussion, LGTM. I'll circle back once a direction has been decided-upon there :)

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 19 above)
Comment (dexonsmith) [reply llvm/lib/Support/TrieRawHashMap.cpp]: The point is catch misuse, not logic errors, so it's okay for the assertion to be relaxed if there's another reasonable limit.

The limit of 20 bits ensures the root's allocation is under 8MB (1M entries). On some platforms, larger contiguous allocations might fail. An assertion could catch platform-dependent memory failures.

Certainly, I imagine a limit of 29 would be palatable? (500M entries, 4GB allocation for the root node)

In any case, it feels like someone is holding the trie "wrong" for root nodes this big. Going bigger is certainly valid in some sense, but I question the benefit of supporting it. If you really want to optimize for super large data sets, you probably want a file-backed trie instead.

I prefer the original limit of 20, unless someone has a specific argument for something else. (I don't feel strongly about what the limit is, just that there should be one.)

File: llvm/include/llvm/ADT/TrieRawHashMap.h [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+  /// \pre \a destroyImpl() was already called.
+  ~ThreadSafeTrieRawHashMapBase();
+  void destroyImpl(function_ref<void(void *ValueMem)> Destructor);
+
+  ThreadSafeTrieRawHashMapBase(ThreadSafeTrieRawHashMapBase &&RHS);
+
+  // Move assignment can be implemented in a thread-safe way if NumRootBits and
+  // NumSubtrieBits are stored inside the Root.
+  ThreadSafeTrieRawHashMapBase &
+  operator=(ThreadSafeTrieRawHashMapBase &&RHS) = delete;
```
Comment (bogner) [inline llvm/include/llvm/ADT/TrieRawHashMap.h:149]: Is the comment here meant to be a TODO/FIXME, or just general commentary? It's a bit confusing to have a comment that says something could be done but then seeing that the code does not do it. Maybe rewording it like "We don't do X because of Y, but if we needed to then we could do Z" would make it clearer.

File: llvm/include/llvm/ADT/TrieRawHashMap.h [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+private:
+  friend class TrieRawHashMapTestHelper;
+  const unsigned short ContentAllocSize;
+  const unsigned short ContentAllocAlign;
+  const unsigned short ContentOffset;
+  unsigned short NumRootBits;
+  unsigned short NumSubtrieBits;
+  class ImplType;
+  // ImplPtr is owned by ThreadSafeTrieRawHashMapBase and needs to be freed in
+  // destoryImpl.
```
Comment (bogner) [inline llvm/include/llvm/ADT/TrieRawHashMap.h:176]:
```suggestion
  // destroyImpl.
```

File: llvm/lib/Support/TrieHashIndexGenerator.h [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+  // the hash, it can start index generation from this level by calling \c hint
+  // function.
+  size_t hint(unsigned Index, unsigned Bit) {
+    assert(Bit < Bytes.size() * 8);
+    assert(Bit == 0 || (Bit - NumRootBits) % NumSubtrieBits == 0);
+    StartBit = Bit;
+    return Index;
+  }
+
+  // Utility funciton for looking up the index in the trie for an object that
```
Comment (bogner) [inline llvm/lib/Support/TrieHashIndexGenerator.h:75]:
```suggestion
  // Utility function for looking up the index in the trie for an object that
```

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 5 above)
Comment (bogner) [reply llvm/lib/Support/TrieRawHashMap.cpp:122]: A code comment here pointing that out would be good, since it definitely looks like a bug at a glance.

**Review by bogner — APPROVED** [review]
A few editorial notes on the comments, but the implementation looks good to me.

**Comment by cachemeifyoucan** [comment]
Address all the review feedback. Also squash the commits manually and rewrite the commit message.

File: llvm/lib/Support/TrieRawHashMap.cpp (hunk 5 above)
Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:123]: This is switch to just `new` to avoid the confusion and overwrite both `new` and `delete` operator. This should align with how other trailing objects are handled.

**Comment by cachemeifyoucan** [comment]
@bogner Let me know what you think of the new commit message to show clear intent for what the data structure is for. I am preparing the follow up PRs for the rest of the implementation and they should be coming soon.

**Comment by bogner** [comment]
> @bogner Let me know what you think of the new commit message to show clear intent for what the data structure is for. I am preparing the follow up PRs for the rest of the implementation and they should be coming soon.

Looks good. Thanks!

File: llvm/lib/Support/TrieRawHashMap.cpp [hunk 26]
```diff
@@ -0,0 +1,515 @@
...
+unsigned ThreadSafeTrieRawHashMapBase::getNumSlotUsed(
+    ThreadSafeTrieRawHashMapBase::PointerBase P) const {
+  assert(!P.isHint() && "Not a valid trie");
+  if (!P.P)
+    return 0;
+  auto *S = dyn_cast<TrieSubtrie>((TrieNode *)P.P);
+  if (!S)
+    return 0;
+  unsigned Num = 0;
+  for (unsigned I = 0, E = S->size(); I < E; ++I)
```
Comment (kstoimenov) [inline llvm/lib/Support/TrieRawHashMap.cpp:426]: This is a causing an error in Sanitizer builds. It is also confusing because you have E defined in the for loop and the if statement. What the intent here?

Comment (cachemeifyoucan) [reply llvm/lib/Support/TrieRawHashMap.cpp:426]: This is just an unused variable. Fix: https://github.com/llvm/llvm-project/pull/114114

I don't understand the other error from https://lab.llvm.org/buildbot/#/builders/168/builds/4907 though as that is a CMake error that I didn't touch.
